cores/arduino/stm32/dma.c

```diff
@@ -0,0 +1,315 @@
+  DmaHandleTab[indice].Init.Priority            = DMA_PRIORITY_MEDIUM;
+  #if defined(STM32L0xx) || defined(STM32L4xx)
+  DmaHandleTab[indice].Init.Request             = DMA_REQUEST_0;
+  #endif
+  
+  #if defined(STM32F4xx)
+  #ifdef ADC1_BASE
+  if (hadc->Instance == ADC1)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_0;
+  }
+  #endif
+  #ifdef ADC2_BASE
+  if (hadc->Instance == ADC2)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_1;
+  }
+  #endif
+  #ifdef ADC3_BASE
+  if (hadc->Instance == ADC3)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_2;
+  }
+  #endif
+  DmaHandleTab[indice].Init.FIFOMode              = DMA_FIFOMODE_DISABLE;
+  DmaHandleTab[indice].Init.FIFOThreshold         = DMA_FIFO_THRESHOLD_HALFFULL;
+  DmaHandleTab[indice].Init.MemBurst              = DMA_MBURST_SINGLE;
+  DmaHandleTab[indice].Init.PeriphBurst           = DMA_PBURST_SINGLE;
+  #endif //STM32F4xx
+  
+  /* Deinitialize  & Initialize the DMA for new transfer */
+  HAL_DMA_DeInit(&DmaHandleTab[indice]);
+  HAL_DMA_Init(&DmaHandleTab[indice]);
+  
+  /* Associate the DMA handle */
+  __HAL_LINKDMA(hadc, DMA_Handle, DmaHandleTab[indice]);
+
+  /* Set clock as active */
+  channelNumber = pinmap_find_dmaChannelNumber(hadc->Instance, DmaMap_DMA);
+  dmaNumber     = pinmap_find_dma(hadc->Instance, DmaMap_DMA);
+  dmaTab[dmaNumber-1][channelNumber] = true;
+
+  /* NVIC configuration for DMA Input data interrupt */
+  irqHandler = pinmap_find_dmaIrqHandler(hadc->Instance, DmaMap_DMA);
+  HAL_NVIC_SetPriority(irqHandler, 1, 0);
+  HAL_NVIC_EnableIRQ(irqHandler);
+}
+
+/**
+  * @brief  This function will deinit the dma and channel corresponding 
+  *         to the ADC peripheral
+  * @param  hadc : the ADC peripheral to use
+  * @retval none
+  */
+void dma_deinit_adc(ADC_HandleTypeDef *hadc)
+{
+  IRQn_Type           irqHandler;
+  int8_t              channelNumber;
+  int8_t              dmaNumber;
+  
+  irqHandler = pinmap_find_dmaIrqHandler(hadc->Instance, DmaMap_DMA);
+  HAL_NVIC_DisableIRQ(irqHandler);
+  HAL_DMA_DeInit(hadc->DMA_Handle);
+  
+  // Disable the clock only if not another channel of the dma is use
+  channelNumber = pinmap_find_dmaChannelNumber(hadc->Instance, DmaMap_DMA);
+  dmaNumber     = pinmap_find_dma(hadc->Instance, DmaMap_DMA);
+  dmaTab[dmaNumber-1][channelNumber] = false;
+  
+ for (int i=0; i<CHANNEL_NUMBER; i++)
+  {
+    if (dmaTab[dmaNumber][i] == true)
+    {
+      break;
+    }
+    if (i == CHANNEL_NUMBER-1)
+    {
+      // All dma channel are inactive
+      if (dmaNumber == 1)
+      {
+      #ifdef __HAL_RCC_DMA1_CLK_ENABLE
+        __HAL_RCC_DMA1_CLK_DISABLE();
+      #endif
+      }
+      else if (dmaNumber == 2)
+      {
+        #ifdef __HAL_RCC_DMA2_CLK_ENABLE
+        __HAL_RCC_DMA2_CLK_DISABLE();
+        #endif
+      }
+    }
+  }
+}
+
+
+/**
+  * @brief  DMA IRQ Handle 
+  * @param  None
+  * @retval None
+  */
+
+#ifdef ADC1_BASE
```

Comment:
Wouldn't it be better to keep generic name ?
DMA1_Channel1_IRQHandler

Reply:
We can't keep generic name because the are different from one device to another. For example for F3 family this is DMA2_Channel1_IRQHandler and for F4 family this is DMA2_Stream1_IRQHandler.  (Channel/Stream)

Reply:
Ok I understand that - I just mean is it really needed that the name contains 'ADC'
it could be always DMA2_Stream1_IRQHandler anyway and redefined only for few families.

Reply:
Ok done in next release.